File: src/main/kotlin/com/xs0/gqlktx/types/gql/GListType.java
package com.xs0.gqlktx.types.gql;

import com.xs0.gqlktx.QueryException;
import com.xs0.gqlktx.schema.builder.TypeKind;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

public class GListType extends GWrappingType {
    public GListType(GType innerType) {
        super(innerType);
    }

    @Override
    public TypeKind getKind() {
        return TypeKind.LIST;
    }

    @Override
    public String getGqlTypeString() {
        return "[" + getWrappedType().getGqlTypeString() + "]";
    }

    @Override
    public void coerceValue(JsonObject raw, String key, JsonObject out) throws QueryException {
        JsonArray rawList;
        try {
            rawList = raw.getJsonArray(key);
        } catch (ClassCastException e) {
            throw new QueryException("List type " + getGqlTypeString() + " needs a list value, but had something else");
        }

        if (rawList == null) {
            out.putNull(key);
        } else {
            JsonArray coerced = new JsonArray();
            for (int i = 0, n = rawList.size(); i < n; i++) {
                getWrappedType().coerceValue(rawList, i, coerced);
            }
            out.put(key, coerced);
        }
    }

    @Override
    public void coerceValue(JsonArray raw, int index, JsonArray out) throws QueryException {
        JsonArray rawList;
        try {
            rawList = raw.getJsonArray(index);
        } catch (ClassCastException e) {
            throw new QueryException("List type " + getGqlTypeString() + " needs a list value, but had something else");
        }

        if (rawList == null) {
            out.addNull();
        } else {
            JsonArray coerced = new JsonArray();
            for (int i = 0, n = rawList.size(); i < n; i++) {
                getWrappedType().coerceValue(rawList, i, coerced);
            }
            out.add(coerced);
        }
    }
}

File: src/main/kotlin/com/xs0/gqlktx/types/gql/GListType.kt
package com.xs0.gqlktx.types.gql

import com.xs0.gqlktx.QueryException
import com.xs0.gqlktx.dom.Value
import com.xs0.gqlktx.dom.ValueList
import com.xs0.gqlktx.dom.ValueNull
import com.xs0.gqlktx.dom.ValueOrNull
import com.xs0.gqlktx.schema.builder.TypeKind

class GListType(innerType: GType) : GWrappingType(innerType) {

    override val kind: TypeKind
        get() = TypeKind.LIST

    override val gqlTypeString: String
        get() = "[" + wrappedType.gqlTypeString + "]"

    override fun coerceValue(raw: Value): ValueList {
        if (raw !is ValueList)
            throw QueryException("List type $gqlTypeString needs a list value, but had something else")

        val coerced = ArrayList<ValueOrNull>()

        for (i in 0 until raw.elements.size) {
            val rawEl = raw.elements[i]

            if (rawEl is ValueNull) {
                if (wrappedType.kind == TypeKind.NON_NULL)
                    throw QueryException("Null value not allowed in this list")
                coerced.add(rawEl)
            } else {
                coerced.add(wrappedType.coerceValue(rawEl as Value))
            }
        }

        return ValueList(coerced)
    }
}
